server.js

```diff
@@ -1,20 +1,119 @@
 const express = require("express");
 const cors = require("cors");
-
+const moment = require("moment");
+var validator = require("email-validator");
 const app = express();
-
 app.use(express.json());
 app.use(cors());
+moment().format();
+
+// TODO add your routes and helper functions here
 
+app.get("/bookings/search", function (request, response) {
+  const queryDate = request.query.date;
+  const term = request.query.term;
+  if (!term && !queryDate) {
+    response.status(400).json("There isn't any match");
+    return;
+  }
+
+  if (moment(queryDate, "YYYY-MM-DD", true).isValid()) {
+    let result = bookings.filter(
+      (booking) =>
+        moment(queryDate).isBetween(
+          booking.checkInDate,
+          booking.checkOutDate
+        ) ||
+        moment(queryDate).isSame(booking.checkInDate) ||
+        moment(queryDate).isSame(booking.checkOutDate)
+    );
+    result.length > 0
+      ? response.send(result)
+      : response.send("Not found a booking ton this date");
+    return;
+  } else if (queryDate) {
+    response
+      .status(400)
+      .json('Please enter a valid date, format should like "YYYY-MM-DD"');
+    return;
+  }
+
+  const searchTerm = bookings.filter((booking) =>
+    `${booking.firstName} ${booking.surname} ${booking.email}`
+      .toLowerCase()
+      .includes(term.toLowerCase())
+  );
+  if (term && searchTerm.length > 0) {
+    response.json(searchTerm);
+  } else if (term) {
+    response.status(400).json("Sorry we couldn't find your term");
+  }
+});
 //Use this array as your (in-memory) data store.
 const bookings = require("./bookings.json");
-
 app.get("/", function (request, response) {
   response.send("Hotel booking server.  Ask for /bookings, etc.");
 });
+//
+//read all bookings
+app.get("/bookings", (request, response) => {
+  response.json(bookings);
+});
 
-// TODO add your routes and helper functions here
+app.post("/bookings", (request, response) => {
+  if (!validator.validate(request.body.email)) {
+    response.status(404).json("Please enter a valid email address");
+    return;
+  }
+
+  let newBooking = {
+    id: bookings.length + 1,
```

Comment:
Relying on the `bookings` length to generate new id may be quite risky. What if you add a new booking, then delete a different one, and then create a new booking again? Two bookings will end up having the same id. I think it would be better to increment last booking's id instead.